[c10d] Land CudaEventCache with roll out flags

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133727

@zdevito added a cache for CudaEvent in https://github.com/pytorch/pytorch/pull/122732. And we want to productionize it with a flag in this PR.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o @xmfan

fduwjj committed:
[c10d] Land CudaEventCache with roll out flags

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Land CudaEventCache with roll out flags"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -481,10 +482,20 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(
+      ncclStartEvent_ = std::make_shared<at::cuda::CUDAEvent>(cudaEventDefault);
+    }
+  }
+  if (cudaEventCacheEnabled) {
```
Is that possible to make the 4 execution branches clearer using the nested if/else?

fduwjj committed:
Update on "[c10d] Land CudaEventCache with roll out flags"


zdevito added a cache for CudaEvent in https://github.com/pytorch/pytorch/pull/122732. And we want to productionize it with a flag in this PR.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -470,6 +470,7 @@ ProcessGroupNCCL::WorkNCCL::WorkNCCL(
     const std::optional<std::vector<at::Tensor>>& inputs,
     bool desyncDebug,
     bool enableTiming,
+    bool cudaEventCacheEnabled,
```
nit: maybe we can make `WorkNCCL` a friend of `ProcessGroupNCCL` so that it can access `ProcessGroupNCCL. cudaEventCacheEnabled_` and we don't have to pass every flag to the constructor of `WorkNCCL`?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -752,6 +761,35 @@ void ProcessGroupNCCL::WorkNCCL::abort() {
 
+ProcessGroupNCCL::CUDAEventCache::CUDAEventCache() {}
+
+std::shared_ptr<at::cuda::CUDAEvent> ProcessGroupNCCL::CUDAEventCache::create(
+    bool timing) {
+  auto deleter = [this, timing](at::cuda::CUDAEvent* event) {
+    std::lock_guard<std::mutex> lock(this->cacheMutex_);
+    this->eventsArray_[timing ? 1 : 0].push_back(event);
+  };
+  at::cuda::CUDAEvent* event = nullptr;
+  {
+    std::lock_guard<std::mutex> lock(cacheMutex_);
+    auto events = eventsArray_[timing ? 1 : 0];
+    if (!events.empty()) {
+      event = events.back();
+      events.pop_back();
+    }
+  }
+  if (!event) {
+    event = new at::cuda::CUDAEvent(
+        timing ? cudaEventDefault : cudaEventDisableTiming);
+  }
+  return std::shared_ptr<at::cuda::CUDAEvent>(event, std::move(deleter));
+}
+
+ProcessGroupNCCL::CUDAEventCache& ProcessGroupNCCL::CUDAEventCache::get() {
+  static ProcessGroupNCCL::CUDAEventCache cache;
+  return cache;
+}
+
```
nit: can you add some comments for this block of code? Thanks!

kwen2501 reviewed (commented):
LGTM. Added two minor comments.
